File: src/main/java/com/github/jonross/jmiser/graph/Dominators.java
/*
 * Copyright (c) 2012 by Jonathan Ross (jonross@alum.mit.edu)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.jonross.jmiser.graph;

import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.util.List;

import com.github.jonross.jmiser.IntLists;
import com.github.jonross.jmiser.Settings;
import com.google.common.collect.Lists;

/**
 * Off-heap implementation of Lengauer-Tarjan for dominators in an {@link IntGraph}.
 * Based on "Computing Dominators and Dominance Frontiers" by Briggs / Harvey.
 * http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.56.8903
 */

public class Dominators {

    private final Settings SETTINGS = new Settings();
    private final List<ByteBuffer> buffers = Lists.newArrayList();
    
    private final IntGraph g;
    private final int max;
    private final IntBuffer ord, rev, parent, semi, idom, ancestor, best;
    private final IntLists buck;
    private int num = 1;

    public Dominators(IntGraph g) {

        this.g = g;
        max = g.maxNode();
        ord = getInts(max+1);
        rev = getInts(max+1);
        parent = getInts(max+1);
        semi = getInts(max+1);
        idom = getInts(max+1);
        ancestor = getInts(max+1);
        best = getInts(max+1);
        buck = new IntLists(new Settings());

        buck.add(max, 0);
        buck.clear(max);

        // step 1
        dfs(1, 0);
        for (int v = 1; v <= max; v++) {
            semi.put(v, v);
            idom.put(v, 0);
            ancestor.put(v, 0);
            best.put(v, v);
        }

        for (int w = max; w > 1; w--) {
            int p = parent.get(w);

            // step 2
            for (long cur = g.walkInEdges(rev.get(w)); cur != 0; cur = g.nextInEdge(cur)) {
                int v = ord.get((int) (cur & 0xFFFFFFFF));
                int u = eval(v);
                int semi_u = semi.get(u);
                if (semi.get(w) > semi_u)
                    semi.put(w, semi_u);
                buck.add(semi.get(w), w);
                link(p, w);
            }

            // step 3
            for (long cur = buck.walk(p); cur != 0; cur = buck.next(cur)) {
                int v = (int) (cur & 0xFFFFFFFF);
                int u = eval(v);
                idom.put(v, semi.get(u) < semi.get(v) ? u : p);
            }
            buck.clear(p);
        }

        // step 4
        idom.put(1, 0);
        for (int w = 2; w <= max; w++)
            if (idom.get(w) != semi.get(w))
                idom.put(w, idom.get(idom.get(w)));
    }

    public int[] get() {
        int[] d = new int[max+1];
        d[0] = d[1] = 0;
        for (int i = 2; i <= max; i++)
            d[i] = rev.get(idom.get(ord.get(i)));
        return d;
    }
    
    private IntBuffer getInts(int size) {
        ByteBuffer buf = SETTINGS.alloc(size * 4);
        buffers.add(buf);
        return buf.asIntBuffer();
    }
    
    public void destroy() {
        SETTINGS.free(buffers.toArray(new ByteBuffer[0]));
        buck.destroy();
    }
    
    private void dfs(int _v, int _p) {
        if (ord.get(_v) == 0) {
            int v = num++;
            ord.put(_v, v);
            rev.put(v, _v);
            parent.put(v, ord.get(_p));
            for (long cur = g.walkOutEdges(_v); cur != 0; cur = g.nextOutEdge(cur))
                dfs((int) (cur & 0xFFFFFFFF), _v);
        }
    }

    private void link(int v, int w) {
        ancestor.put(w, v);
    }

    private int eval(int v) {
        if (ancestor.get(v) != 0)
            compress(v);
        return best.get(v);
    }

    private void compress(int v) {
        int a = ancestor.get(v);
        if (ancestor.get(a) == 0)
            return;
        compress(a);
        if (semi.get(best.get(v)) > semi.get(best.get(a)))
            best.put(v, best.get(a));
        ancestor.put(v, ancestor.get(a));
    }

}

File: src/main/scala/trilby/graph/Dominators.scala
/*
 * Copyright (c) 2013 by Jonathan Ross (jonross@alum.mit.edu)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package trilby.graph

import java.nio.ByteBuffer
import java.nio.IntBuffer
import trilby.nonheap.NHUtils
import trilby.nonheap.HugeArray
import trilby.util.Oddments._

/**
 * Off-heap implementation of Lengauer-Tarjan for dominators in an {@link IntGraph}.
 * Based on their paper and also "Computing Dominators and Dominance Frontiers" by 
 * Briggs / Harvey.
 * 
 * http://www.cc.gatech.edu/~harrold/6340/cs6340_fall2010/Readings/lengauer91jul.pdf
 * http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.56.8903
 */

class Dominators(val g: IntGraph, onHeap: Boolean) {

    var buffers: List[ByteBuffer] = Nil
    var dmax = 0
    val PAR = 64
    
    private[this] def getInts(size: Int) = {
        val buf = NHUtils.alloc(size * 4, false)
        buffers = buf :: buffers
        buf.asIntBuffer
    }
    
    val ord = getInts(g.maxNode + 1)
    val parent = getInts(g.maxNode + 1)
    val rev = getInts(g.maxNode + 1)
    
    // dfs(1, 0)
    dfs()
    
    val semi = getInts(dmax + 1)
    val idom = getInts(dmax + 1)
    val ancestor = getInts(dmax + 1)
    val best = getInts(dmax + 1)
    val buck = new IntLists(onHeap)
    
    init()
    
    def init() {

        for (offset <- 1 to PAR par) {
            for (v <- offset to dmax by PAR) {
                rev.put(ord.get(v), v)
                semi.put(v, v)
                idom.put(v, 0)
                ancestor.put(v, 0);
                best.put(v, v)
            }
        }

        for (w <- dmax until 1 by -1) {
            val p = parent.get(w)

            // step 2
            var cur = g.walkInEdges(rev.get(w))
            while (cur.valid) {
                val v = ord.get(cur.value)
                if (v > 0) {
                    val u = eval(v)
                    val semi_u = semi.get(u)
                    if (semi.get(w) > semi_u)
                        semi.put(w, semi_u)
                    buck.add(semi.get(w), w)
                    link(p, w)
                }
                cur = g.nextInEdge(cur)
            }

            // step 3
            var bcur = buck.walk(p)
            while (bcur.valid) {
                val v = bcur.value
                val u = eval(v)
                idom.put(v, if (semi.get(u) < semi.get(v)) u else p)
                bcur = buck.next(bcur)
            }
            buck.clear(p)
        }

        // step 4
        idom.put(1, 0)
        for (w <- 2 to dmax) {
            if (idom.get(w) != semi.get(w)) {
                idom.put(w, idom.get(idom.get(w)))
            }
        }
    }

    def get() = {
        val d = new HugeArray.OfInt(g.maxNode + 1, onHeap)
        for (offset <- 1 to PAR par) {
            for (v <- (offset+1) to dmax by PAR) {
                d(rev.get(v)) = rev.get(idom.get(v))
            }
        }
        d
    }
    
    def free() {
        buffers.foreach(NHUtils.free)
        buck.free()
    }
    
    // For reasons TBD, this DFS creates a subtle difference in the resulting tree for
    // large graphs.  The DFS walks the child list in reverse order because of the way
    // CompactIntGraph edges are filled.  It should not matter to the results, but does.
    // To investigate.
    
    private def dfs() {
        new PreorderDFS {
            def maxNode = g.maxNode
            def visit(_v: Int) {
                // get at parent by pushing it ahead of target node
                val _p = stack.pop()
                dmax += 1
                val v = dmax
                ord.put(_v, v)
                // rev.put(v, _v)
                parent.put(v, ord.get(_p))
            }
            def addChildren(_v: Int) {
                var cur = g.walkOutEdges(_v)
                while (cur.valid) {
                    add(_v, cur.value)
                    cur = g.nextOutEdge(cur)
                }
            }
            stack.push(0)
            add(1)
        }.run()
    }
    
    private[this] def link(v: Int, w: Int) {
        ancestor.put(w, v)
    }

    private[this] def eval(v: Int) = {
        if (ancestor.get(v) != 0) {
            compress(v)
        }
        best.get(v)
    }

    private[this] def compress(v: Int) {
        val a = ancestor.get(v)
        if (ancestor.get(a) == 0) {
            return
        }
        compress(a)
        if (semi.get(best.get(v)) > semi.get(best.get(a)))
            best.put(v, best.get(a))
        ancestor.put(v, ancestor.get(a))
    }

}
